cirq-core/cirq/ops/permutation_gate.py

```diff
@@ -74,6 +74,15 @@ def num_qubits(self):
     def _has_unitary_(self):
         return True
 
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        qubit_ids = [*range(len(qubits))]
+        for i in range(len(qubits)):
+            q_i = qubit_ids.index(self._permutation.index(i))
+            for j in range(q_i, i, -1):
+                yield swap_gates.SWAP(qubits[j], qubits[j - 1])
```

Comment:
Note that the decomposition here is not intended to be architecture aware. The goal here is to simply write some decomposition in terms of other "simpler" gates which eventually ends up decomposing into the native gateset. 

For device specific routing, we would need more sophisticated transformers which take care of device topology, but that is out of scope of this PR. The choice to use adjacent index swaps which can be useful for a linear qubit layout is an arbitrary one, mostly because running circuits on a chain of qubits is a common use case and a low hanging fruit.

Reply:
If this is meant to be architecture-agnostic, why don't we swap qubits directly (ignoring adjacency)?

Reply:
As I said, it's a choice we need to make. Both the approaches would be valid for the scope of this PR. 

If a user is relying on the default decomposition of Cirq for compiling their circuit, it means they have already given up fine grained control. The reason I want to stick to this approach is that it's a low hanging fruit which targets a common use case of running circuits on a chain of qubits. It's not going to be enough for every scenario, but still slightly better than assuming all to all connectivity.

Reply:
> It's not going to be enough for every scenario, but still slightly better than assuming all to all connectivity.

I'm not convinced that linear-swap is better than all-to-all, or even equally valid. Compared to linear-swap, all-to-all is:
- cheaper to run on simulators (which use default decomposition)
- easier to convert to efficient circuits in device-specific layout
- cheaper to construct here (low priority, but worth considering)

I'd support adding a separate method for converting to linear-swap, but the default should be all-to-all.

Reply:
> cheaper to run on simulators (which use default decomposition)

The class already has a `_apply_unitary_` defined, which should be used by simulators over the default decomposition. Is that not the case?

Reply:
This is true for Cirq-internal simulators, but not in general. For example, qsim attempts to decompose gates to a specific subset that it natively supports, falling back on conversion to unitary only if no decomposition is defined ([code](https://github.com/quantumlib/qsim/blob/6443786bd1f283cbfa8d25151100a9ae640e3807/qsimcirq/qsim_circuit.py#L317)).

Reply:
Opened https://github.com/quantumlib/Cirq/issues/5097 to track the optimization.